### PR TITLE
nvidia_x11: 390.87 -> 410.66

### PR DESCRIPTION
**Pull request by baracoder**
###### Motivation for this change

Nvidia stable branch driver was updated to [410.66](https://www.nvidia.com/object/unix.html). With the work from #45030 this change is simple.

**Note:** Nvidia dropped support for x86. There are legacy versions present, but the last driver for x86 is 390.87. So maybe I should add it as a separate version like the legacy versions below?

/cc @Infinisil, @eadwu, @tenten8401
see also #44461

###### Things done

- [x] Added x86 switch for latest supported driver
- [x] Removed beta
- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Infinisil** [comment]
> So maybe I should add it as a separate version like the legacy versions below?

Yeah that sounds like a good idea

**Comment by eadwu** [comment]
Never really did this in my PR since I assumed most people who would use it probably lie on amd64, though the assignment in `videoDrivers` should probably be asserted (at least for nvidia) and defaulted to 3xx for 32bit.

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/hardware/video/nvidia.nix#L14

**Comment by cpages** [comment]
I understand that v410 is now considered stable, so I would do 2 things:
- Choose a different version for the stable attribute depending on the platform (keep 390.87 for x86 users)
- Remove the beta version, since I understand there's no beta currently (be aware that there are references in the nvidia module and in all-packages)

In this way we keep a working default for both platforms

**Comment by baracoder** [comment]
I have updated the commit:

Some packages reference the default `nvidia_x11` package directly. 
e.g. https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/all-packages.nix#L921
and https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/all-packages.nix#L11493

So `nvidia_x11` should evaluate to the latest stable driver package for the platform I think.

File: pkgs/os-specific/linux/nvidia-x11/default.nix [hunk 1]
```diff
@@ -16,7 +16,19 @@ let
...
-  stable = generic {
+  stable = if stdenv.hostPlatform.system == "x86_64-linux" then stable_410 else stable_390;
+
+  stable_410 = generic {
+    version = "410.66";
+    sha256_64bit = "05xjzvj0fgmkpz36dbd7hy2vzl6xxiflzx7kml3k7ad9gy2svdlg";
+    settingsSha256 = "1nsxz1byshgjs3c03lyx6ya36dp0f2vg2l0d9pkh1i6cpzkp53kz";
+    persistencedSha256 = "0m4wdpb8w4y323d8py105p9hizwmf2ai8frkl7h77sn3ski17zw6";
+
+    patches = lib.optional (kernel.meta.branch == "4.19") ./drm_mode_connector.patch;
```
Comment (eadwu) [inline pkgs/os-specific/linux/nvidia-x11/default.nix]: On 4xx this patch isn't needed anymore.

**Review by jokogr — APPROVED** [review]

**Comment by baracoder** [comment]
@cpages I have set `nvidia_x11_beta` to be the same as `nvidia_x11` and added a switch for `nvidia_x11` depending on host architecture

**Review by infinisil — APPROVED** [review]
This looks good to me. It might be nice to have some people with different graphics cards test this.

**Review by cpages — APPROVED** [review]
@Infinisil I don't think this new version even drops support for any older card, so for me this can be pushed

**Review by jokogr — APPROVED** [review]

**Comment by jokogr** [comment]
The update works just fine with my 970.

**Comment by sib-null** [comment]
This merge broke my system. I have a 64-bit system with an Nvidia GeForce GTX 560 Ti graphics card, which is only supported in the now legacy 390.x series. According to the changes, because I have a 64-bit system, the 410.x series would be used instead of the 390.x. However, there are no legacy_390 options to choose from, so I am stuck without a proper driver for the moment.

Please see #49679

**Comment by ben0x539** [comment]
Can we revert this until the 390.x series is packaged separately? Being stuck on an old revision of nixos-unstable is no fun. :(

**Comment by jokogr** [comment]
@ben0x539 if I were you, I would test #51294 and try to merge this instead of reverting this one.

**Comment by ben0x539** [comment]
Hm, that PR hasn't even been reviewed yet, it seems like it getting merged and then making its way to nixos-unstable might be a while off. If I understand correctly, it also depends on #49703 to function as a workaround, which seems stuck. The prospect of all of that needing to land before I can update my system again is somewhat discouraging.

**Comment by eadwu** [comment]
#51294 doesn't depend on #49703 though it should still work even if you merge both. Just adding `nvidiaLegacy390` to `videoDrivers` should work just fine.

**Comment by sib-null** [comment]
@eadwu, can you explain why you prefer #49703? It seems that with #49703 you won't get automatic upgrades anymore?

**Comment by eadwu** [comment]
With #51294, every time there needs to be a legacy package, the way most users use it would be through `videoDrivers`, so someone would have to add it there every single time. Though through #49703, you just need add `nvidia` to `videoDrivers` and then set `hardware.nvidia.package` to the desired package and it should give the same results. Also keeps the nvidia packages in its `default.nix` since the exposure is through `config.boot.kernelPackages` instead of searching and matching through `videoDrivers`.

**Pull request merged**